src/custom/components/swap/EthWethWrap/index.tsx

```diff
@@ -113,160 +67,600 @@ const ButtonWrapper = styled.div`
-        {loading ? <Loader /> : <Trans>Wrap my {nativeSymbol}</Trans>}
-      </ButtonPrimary>
+      <SimpleAccountDetails
+        pendingTransactions={Object.values(pendingHashMap)}
+        confirmedTransactions={[]}
+        $margin="12px 0 0"
+      />
+      {renderActionButton()}
     </Wrapper>
   )
 }
+
+type DerivedEthFlowStateProps = Pick<ModalTextContentProps, 'wrapSubmitted' | 'approveSubmitted' | 'isExpertMode'> &
+  Pick<Props, 'needsApproval' | 'needsWrap'> & {
+    approveError: Error | null
+    wrapError: Error | null
+    approveSubmitted: boolean
+    wrapSubmitted: boolean
+    isApprovePending: boolean
+    isWrapPending: boolean
+  }
+
+// returns derived ethflow state from current props
+function _getDerivedEthFlowState(params: DerivedEthFlowStateProps) {
+  const {
+    approveError,
+    wrapError,
+    approveSubmitted,
+    wrapSubmitted,
+    isApprovePending,
+    isWrapPending,
+    needsApproval,
+    needsWrap,
+    isExpertMode,
+  } = params
+  // approve state
+  const approveSentAndSuccessful = !approveError && approveSubmitted && !isApprovePending
+  const approveFinished = !needsApproval || approveSentAndSuccessful
+  // wrap state
+  const wrapSentAndSuccessful = !wrapError && wrapSubmitted && !isWrapPending
+  const wrapFinished = !needsWrap || wrapSentAndSuccessful
+
+  let state = EthFlowState.Loading
+  if (approveFinished && wrapFinished) {
+    state = EthFlowState.SwapReady
+  } else if (isExpertMode && needsWrap && !wrapSentAndSuccessful && needsApproval && !approveSentAndSuccessful) {
+    state = EthFlowState.WrapAndApproveNeeded
+  } else if (needsWrap && !wrapSentAndSuccessful) {
+    state = EthFlowState.WrapNeeded
+  } else if (needsApproval && !approveSentAndSuccessful) {
+    state = EthFlowState.ApproveNeeded
+  } else {
+    state = EthFlowState.Loading
+  }
+
+  return state
+}
+
+type ModalTextContentProps = {
+  wrappedSymbol: string
+  nativeSymbol: string
+  state: EthFlowState
+  isExpertMode: boolean
+  isNative: boolean
+  wrapSubmitted: boolean
+  approveSubmitted: boolean
+}
+
+// returns modal content: header and descriptions based on state
+function _getModalTextContent(params: ModalTextContentProps) {
+  const { wrappedSymbol, nativeSymbol, state, isExpertMode, isNative, wrapSubmitted, approveSubmitted } = params
+  // common text
+  const swapHeader = `Swap ${wrappedSymbol}`
+  const swapDescription = `Click "Swap" to submit an off-chain transaction and swap your ${wrappedSymbol}`
+  const wrapHeader = isNative ? `Wrap your ${nativeSymbol}` : `Unwrap your ${wrappedSymbol}`
+
+  let header = '',
+    description: string | null = null
+  // in expert mode we start the logic right away
+  // and sometimes there can be 2 simultaneous transactions
+  if (isExpertMode) {
+    if (state === EthFlowState.WrapAndApproveNeeded) {
+      header = `${isNative ? 'Wrap ' + nativeSymbol : 'Unwrap ' + wrappedSymbol} and Approve`
```

Comment:
Does "unwrapping and approve" make sense? In which scenario would a user want to do it?

Reply:
woudn't happen, ure right